File: Backpack/Label/Classes/BPKLabel.m
/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2018-2021 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#import "BPKLabel.h"

#import <Backpack/Color.h>
#import <Backpack/Common.h>

#import "BPKTextDefinition.h"

NS_ASSUME_NONNULL_BEGIN
@interface BPKLabel ()

- (void)setupWithStyle:(BPKFontStyle)style;
@property(strong, nonatomic) NSMutableArray<BPKTextDefinition *> *persistentStyleRanges;

@end

@implementation BPKLabel

- (instancetype)initWithFontStyle:(BPKFontStyle)style {
    BPKAssertMainThread();
    self = [super initWithFrame:CGRectZero];

    if (self) {
        [self setupWithStyle:style];
    }

    return self;
}

- (nullable instancetype)initWithCoder:(NSCoder *)aDecoder {
    BPKAssertMainThread();
    self = [super initWithCoder:aDecoder];

    if (self) {
        [self setupWithStyle:BPKFontStyleTextBase];
    }

    return self;
}

- (instancetype)initWithFrame:(CGRect)frame {
    BPKAssertMainThread();
    self = [super initWithFrame:frame];

    if (self) {
        [self setupWithStyle:BPKFontStyleTextBase];
    }

    return self;
}

- (void)updateTextStyle {
    if (self.text == nil) {
        self.attributedText = nil;
        return;
    }

    NSMutableAttributedString *newAttributedString = [[BPKFont attributedStringWithFontStyle:self.fontStyle andColor:self.textColor onAttributedString:self.attributedText] mutableCopy];

    // Recreate the attributed string from the persisted definitions
    for (BPKTextDefinition *styleRange in self.persistentStyleRanges) {
        NSDictionary<NSAttributedStringKey, id> *newAttributes = [self getAttributesWithFontStyle:styleRange.fontStyle];
        [newAttributedString setAttributes:newAttributes range:styleRange.range];
    }

    super.attributedText = newAttributedString;
}

- (NSDictionary<NSAttributedStringKey, id> *)getAttributesWithFontStyle:(BPKFontStyle)fontStyle {
    if (self.textColor) {
        NSDictionary<NSAttributedStringKey, id> *customAttributes = @{NSForegroundColorAttributeName: self.textColor};
        return [BPKFont attributesForFontStyle:fontStyle withCustomAttributes:customAttributes];
    } else {
        return [BPKFont attributesForFontStyle:fontStyle];
    }
}

- (void)setText:(NSString *_Nullable)text {
    BPKAssertMainThread();

    [self.persistentStyleRanges removeAllObjects];

    [super setText:text];
    [self updateTextStyle];
}

- (void)setFontStyle:(BPKFontStyle)fontStyle {
    BPKAssertMainThread();
    _fontStyle = fontStyle;

    self.text = self.attributedText.string;
}

- (void)setFontStyle:(BPKFontStyle)fontStyle range:(NSRange)range {
    [_persistentStyleRanges addObject:[[BPKTextDefinition alloc] initWithFontStyle:fontStyle range:range]];

    NSMutableAttributedString *resultingString = [self.attributedText mutableCopy];
    NSDictionary<NSAttributedStringKey, id> *newAttributes = [self getAttributesWithFontStyle:fontStyle];
    [resultingString setAttributes:newAttributes range:range];

    [self setAttributedText:resultingString];
}

#pragma mark - Private

- (void)setupWithStyle:(BPKFontStyle)style {
    self.persistentStyleRanges = [[NSMutableArray alloc] init];
    self.fontStyle = style;
    self.textColor = BPKColor.textPrimaryColor;
}

@end
NS_ASSUME_NONNULL_END

File: Backpack/Label/Classes/BPKLabel.swift
/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2018 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import UIKit

/// `BPKLabel` is a subclass of `UILabel` which uses the Skyscanner style.
@available(*, deprecated, message: "BPKLabel is deprecated. Use the SwiftUI BPKText instead.")
@objcMembers
@IBDesignable
public final class BPKLabel: UILabel {

    /// The font style used for the label.
    ///
    /// See `BPKFontStyle` for the integer values to use when setting from Interface Builder.
    public var fontStyle: BPKFontStyle = .textBodyDefault {
        didSet {
            guard fontStyle != oldValue else { return }
            text = attributedText?.string
        }
    }

    private var persistentStyleRanges: [BPKTextDefinition] = []

    // MARK: - Initializers

    /// Create a `BPKLabel` with a specific BPKFont style.
    ///
    /// - Parameter fontStyle: Font style to be used by the label.
    public init(fontStyle: BPKFontStyle) {
        super.init(frame: .zero)
        commonInit(fontStyle: fontStyle)
    }

    public required init?(coder: NSCoder) {
        super.init(coder: coder)
        commonInit(fontStyle: .textBodyDefault)
    }

    public override init(frame: CGRect) {
        super.init(frame: frame)
        commonInit(fontStyle: .textBodyDefault)
    }

    // MARK: - Public

    /// Set the font style for a given range of the label.
    ///
    /// - Parameters:
    ///   - fontStyle: The style of the text being applied.
    ///   - range: The range to apply the font style to.
    public func setFontStyle(_ fontStyle: BPKFontStyle, range: NSRange) {
        persistentStyleRanges.append(BPKTextDefinition(fontStyle: fontStyle, range: range))

        guard let currentAttributedText = attributedText else { return }
        let resultingString = NSMutableAttributedString(attributedString: currentAttributedText)
        let newAttributes = attributes(for: fontStyle)
        resultingString.setAttributes(newAttributes, range: range)
        attributedText = resultingString
    }

    // MARK: - Overrides

    public override var text: String? {
        didSet {
            persistentStyleRanges.removeAll()
            applyTextStyle()
        }
    }

    // MARK: - Private

    private func commonInit(fontStyle: BPKFontStyle) {
        self.fontStyle = fontStyle
        super.textColor = BPKColor.textPrimaryColor
        adjustsFontForContentSizeCategory = BPKFontManager.sharedInstance().dynamicTypeEnabled
        applyTextStyle()
    }

    private func applyTextStyle() {
        guard text != nil else {
            attributedText = nil
            return
        }

        guard fontStyle.rawValue > 0 else { return }

        guard let currentAttributedText = attributedText else { return }
        let newAttributedString = NSMutableAttributedString(
            attributedString: BPKFont.attributedString(
                with: fontStyle,
                andColor: textColor,
                on: currentAttributedText
            )
        )

        for styleRange in persistentStyleRanges {
            let newAttributes = attributes(for: styleRange.fontStyle)
            newAttributedString.setAttributes(newAttributes, range: styleRange.range)
        }

        super.attributedText = newAttributedString
    }

    private func attributes(for fontStyle: BPKFontStyle) -> [NSAttributedString.Key: Any] {
        if let textColor {
            let customAttributes: [NSAttributedString.Key: Any] = [.foregroundColor: textColor]
            return BPKFont.makeAttributes(fontStyle: fontStyle, customAttributes: customAttributes)
        } else {
            return BPKFont.makeAttributes(fontStyle: fontStyle)
        }
    }
}
